pps_mw_training/utils/calculate_mean_std.py

```diff
@@ -0,0 +1,38 @@
+from typing import Any
+from pathlib import Path
+import numpy as np  # type: ignore
+import xarray as xr  # type: ignore
+
+
+def get_std_mean(
+    input_files: list[Path], input_params: list[dict[str, Any]]
+) -> list[dict[str, Any]]:
+    """calculate std and mean for the input dataset to normalise"""
+
+    stats = {}
+    with xr.open_dataset(input_files[0]) as ds:
+        for parameter in ds:
+            data = ds[parameter].values
+            stats[parameter] = {
+                "n": np.sum(np.isfinite(data)),
+                "sum": np.nansum(data),
+                "sum_of_squares": np.nansum(data**2),
+            }
+
+    for file in input_files[1:]:
+        with xr.open_dataset(file) as ds:
+            for parameter in ds:
+                data = ds[parameter].values
+                stats[parameter]["n"] += np.sum(np.isfinite(data))
+                stats[parameter]["sum"] += np.nansum(data)
+                stats[parameter]["sum_of_squares"] += np.nansum(data**2)
+
+    for p in input_params:
```

Comment:
would look better if you split this into three  functions.
The first function could be named `get_stats` and include the stuff above thos row.

The second one could be named `get_std_mean -> tuple[float, float]` so you can do these calculations without these long names that almost makes it impossible to understand it.

And the third one control everything, and have the same signature as the the current function.

Reply:
Ok, I also realised, this standalone function was never called in any script. I only used it once to set up the std/mean. So now I added a cli to this to run it. It also helps me to test it.

Reply:
just a tip. you are aware that you can create a dict using a loop and merge two dictionaries with the   `|` operator? This allows for writing oneliners and more readable code. As example what you had in the end could be replaced by something like this (if you had a get_std_mean function):  
```
for file in input_files:
    with xr.open_dataset(file) as ds:
        for parameter in ds:
            data = ds[parameter].values
            stats[parameter]["n"] += np.sum(np.isfinite(data))
            stats[parameter]["sum"] += np.nansum(data)
            stats[parameter]["sum_of_squares"] += np.nansum(data**2)

std_mean = {parameter: get_std_mean(stat) for parameter, stat in stats.items()}
return [param | std_mean[param["name"]] for param in input_params]

# or the two rows above can even be done in one row

return [param | get_std_mean(stats[param["name"]]) for param in input_params]
```

Reply:
> Ok, I also realised, this standalone function was never called in any script. I only used it once to set up the std/mean. So now I added a cli to this to run it. It also helps me to test it.

I do not really understand why we need a cli to this utils function. If you really want this, you should add the cli part to the scripts directory, and only have the "basic" functions in the current utils module. A utils module should in principle not import from e.g. `pps_mw_training.pipelines.cloud_base` as it is now.

Reply:
This function is never used anywhere in the training. I just ran it once to update the mean/std. No need to run it again once they are updated.  I added the cli as I wanted to test the func after the changes the made. But I also do not want a script for this in "scripts".
I understand we should not have imports to cloud_base in utils, so I will remove the cli part.

Reply:
ok. I think it would be enough if we had functionality for getting mean mean and std,  with this signature:

`get_std_mean(input_files: list[Path], params: list[str]) -> list[dict[str, float]]`:

i.e. skip the updating part. If you look at my example a few comments above we can then simplify the code.

Reply:
To be clear what I mean, your function could look something like:
```
def get_std_mean(input_files: list[Path], params: list[str]) -> list[dict[str, float]]:
    """Get standard deviation and mean associated to given parameters of datataset."""

    def _get_std_mean(n: int, sum: float, sum_of_squares: float) -> dict[str, float]:
        """Get standard deviation and mean from intermediate stats."""
        return {
            "mean": sum / n,
            "std": np.sqrt((sum_of_squares / n) - ((sum / n) ** 2)),
        }

    stats = {p: {"n": 0, "sum": 0.0, "sum_of_squares": 0.0} for p in params}
    for file in input_files:
        with xr.open_dataset(file) as ds:
            for p in params:
                stats[p]["n"] += np.count_nonzero(ds[p].values)
                stats[p]["sum"] += np.nansum(ds[p].values)
                stats[p]["sum_of_squares"] += np.nansum(ds[p].values ** 2)

    return {p: _get_std_mean(**stat) for p, stat in stats.items()}
```

Reply:
I just pushed an update to this function. Doesnt look like this but is still better than before

Reply:
ok. just note that you can reduce the complexity / increase readability of the function by changing the input type.

Reply:
I would prefer to have inputs as the format of settings INPUT_PARAMS. Then I can make an easy copy paste if i want to to update the parameters.

Reply:
I am ok with it, although I am not totally happy with a functions that does more than one expect from the name. 

The thing is that this function now includes some obscured logic that make code using it difficult to understand and describe.
For example, I rembember how you used to use it, something like: `input_params = get_std_mean(files, input_params)` , but what this effectively did for you was to update `mean` and `std` of the parameters. But renaming the function to `update_std_mean` is not good either, since an actual update requires that `mean` and `std` are present in the input. Maybe `set_std_mean` would be a better name.

What I mean is that our code will be more understandable and testable if we write simple functions with names that describes what they actually do.